keps/sig-scheduling/20200116-coscheduling-plugin-based-on-scheduler-framework.md

```diff
@@ -0,0 +1,217 @@
+ 
+## Non-Goals
+ Discuss the API definition of `PodGroup`.
+ 
+## Use Cases
+when running a Tensorflow/MPI job, all tasks of a job must be start together; otherwise, did not start anyone of tasks. If the resource is enough to run all 'tasks', everything is fine; but it's not true for most of case, especially in the on-premises environment. In worst case, all jobs are pending here because of deadlock: every job only start part of tasks, and waits for the other tasks to start. In worst case, all jobs are pending leading to a deadlock.
+ 
+## Terms
+
+- **pgPod**: pod belongs to some `PodGroup`.
+- **regularPod**: a regular `Pod` (which doesn't have `PodGropuName` set).
+
+## Proposal
+
+In order to implement coscheduling, we developed plugins in different extension points. In `QueueSort`  we ensure that the Pods belonging to the same PodGroup are queued back-to-back. For example, suppose PodGroup A owns Pod-A1, Pod-A2, Pod-A3, while PodGroup B owns Pod-B1, Pod-B2. The pods of the two PodGroups should not interleave - it should be always <PodGroup-A, PodGroup-B> or the other way around; but never <Pod-A1, Pod-B1, Pod-A2, ...>. In `Permit` phase we put the pod that doesn't meet min-available into the WaitingMap and reserve resources until min-available are met or timeout is triggered. In `Unreserve` phase，clean up the pods that timed-out.
+
+![image](./20200116-coscheduling-plugin-based-on-scheduler-framework-extensions.png)
+
+
+## Design Details
+
+### PodGroup
+
+We use a special label named ```pod-group/name``` to define a PodGroup. Pods that set this label and use the same value belong to the same PodGroup. This is a short term solution, in the future if the definition of `PodGroup` is accepted by the community, we will define it directly through the CRD of `PodGroup`. This is not the focus of this proposal.
+
+```yaml 
+labels:
+     pod-group/name: nginx
+     pod-group/min-available: "2"
+``` 
+
+### Coscheduling
+```go
+// Coscheduling is a plugin that implements the mechanism of gang scheduling.
+type Coscheduling struct {
+    FrameworkHandle     	framework.FrameworkHandle
+    PodLister               corelisters.PodLister
+    // Key is the name of PodGroup.
+    PodGroupInfos       	map[string]PodGroupInfo
+    // Name of the last scheduled podgroup
+    LastPodGroup        	string
+}
+ 
+type PodGroupInfo struct {
+    // LastFailureTimestamp stores the timestamp of last scheduling failure.
+    LastFailureTimestamp 	time.Time
+    UID           			types.UID
+    MinAvailable  		int
+    Name                                 String
+}
+```
+
+1.  `PodGroupInfo` is initialized the first time the pod belongs to the PodGroup is encountered, and LastFailureTimestamp is updated every time the PodGroup fails to schedule.
+2.  `LastPodGroup` records which `PodGroup` the last scheduled pod belongs to.
+3.  `UID` is the unique identification value used to distinguish different podgroups.
+ 
+
+### Extension points
+
+#### QueueSort
```

Comment:
I'd like to discuss implementing `QueueSort` extension points in the Coscheduling plugin.  It is because __only one __ `QueueSort` plugin is allowed in the scheduling framework.    What if some user already has some plugin that implements `QueueSort` plugin? And,  I think we better to document how we support the case after discussion.

Reply:
Yes, we should not dependent on timeout which impact the throughput of scheduler.